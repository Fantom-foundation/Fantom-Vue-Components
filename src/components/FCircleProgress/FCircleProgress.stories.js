import { withA11y } from '@storybook/addon-a11y';

import FCircleProgress from './FCircleProgress.vue';

export default {
    title: 'FCircleProgress',
    component: FCircleProgress,
    decorators: [withA11y],
};

export const Default = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :value="75" />
        </div>
    `,
});

export const ClampValue = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :value="parseFloat(value)" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const ConvertToPercentages = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress convert-to-percentages :value="value" :min="-50" :max="50" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const CanExceed = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress can-exceed :value="value" :min="0" :max="50" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const StrokeWidth = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :stroke-width="10" :value="value" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const Animate = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress animate :value="value" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const Styles = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :styles="styles" convert-to-percentages :value="parseFloat(value)" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 30,
            styles: [
                {
                    value: 0,
                    color: 'red',
                    textStyle: { fill: 'red' },
                },
                {
                    value: 40,
                    color: '#4a5ef5',
                    textStyle: { fill: '#4a5ef5' },
                },
                {
                    value: 80,
                    color: 'green',
                    textStyle: { fill: 'green' },
                },
            ],
        };
    },
});

export const Dots = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :dots="dots" convert-to-percentages :styles="styles" :value="parseFloat(value)" />
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 50,
            dots: [
                {
                    value: 30,
                    color: 'green',
                },
                {
                    value: 75,
                    color: 'red',
                },
            ],
            styles: [
                {
                    value: 30,
                    color: 'green',
                },
                {
                    value: 75,
                    color: 'red',
                },
            ],
        };
    },
});

export const Slot = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :value="value">
                <template #default="{ value }">{{ value }}..</template>
            </f-circle-progress>
            <br />
            <input type="number" v-model="value" class="mat-5" aria-label="value" />
        </div>
    `,
    data() {
        return {
            value: 10,
        };
    },
});

export const Variations = () => ({
    components: { FCircleProgress },
    template: `
        <div>
            <f-circle-progress :value="80" :style="s100" />
        </div>
    `,
    data() {
        return {
            s100: {
                '--fcircleprogress-width': 'calc(100vh - 32px)',
                '--fcircleprogress-height': 'calc(100vh - 32px)',
                '--fcircleprogress-circle-color': 'blueviolet',
                '--fcircleprogress-color': 'blueviolet',
            },
        };
    },
});
