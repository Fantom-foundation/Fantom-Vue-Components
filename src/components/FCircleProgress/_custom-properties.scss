.fcircleprogress {
    --fcircleprogress-width: 152px;
    --fcircleprogress-height: 152px;
    --fcircleprogress-stroke-width: 5;
    --fcircleprogress-transition-duration: 250ms;
    --fcircleprogress-transition-func: ease;
    --fcircleprogress-color: #9299a6;
    //--fcircleprogress-font-size: 0.4;
    --fcircleprogress-font-size: 42px;
    --fcircleprogress-font-weight: bold;
    --fcircleprogress-percentage-font-size: 0.4em;
    --fcircleprogress-stroke-linecap: round;
    --fcircleprogress-circle-color: var(--f-color-primary-5, #1a6aff);
    --fcircleprogress-bg-circle-color: #e6e6e6;
}
