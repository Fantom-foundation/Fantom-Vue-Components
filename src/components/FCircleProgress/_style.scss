@use 'custom-properties';

.fcircleprogress {
    position: relative;
    display: inline-block;
    width: var(--fcircleprogress-width);
    height: var(--fcircleprogress-height);

    svg {
        width: 100%;
        height: 100%;

        /*
        circle {
            r: calc(50 - var(--fcircleprogress-stroke-width) * 0.5);
        }
        */
    }

    &_value {
        fill: var(--fcircleprogress-color);
        font-size: var(--fcircleprogress-font-size);
        font-weight: var(--fcircleprogress-font-weight);
        letter-spacing: -3px;

        &_percentage {
            font-size: var(--fcircleprogress-percentage-font-size);
        }
    }

    &_circle {
        stroke: var(--fcircleprogress-circle-color);
        stroke-linecap: var(--fcircleprogress-stroke-linecap);
    }

    &_bgcircle {
        stroke: var(--fcircleprogress-bg-circle-color);
        stroke-linecap: var(--fcircleprogress-stroke-linecap);
    }

    &-animate {
        .fcircleprogress_circle {
            transition: stroke-dashoffset var(--fcircleprogress-transition-duration)
                var(--fcircleprogress-transition-func);
        }
    }
}
