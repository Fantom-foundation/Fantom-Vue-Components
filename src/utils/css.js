import { isObject } from './index.js';

const lengthAndUnitRE = /([\d.]+)([\w%]*)/;

/**
 * Parse length of css property - like '200px', '50%', ...
 *
 * @param {string} _value
 * @return {{unit: string, len: number}}
 */
export function getLengthAndUnit(_value) {
    const match = typeof _value === 'string' ? lengthAndUnitRE.exec(_value) : null;
    const obj = { len: 0, unit: '' };

    if (match && match.length === 3) {
        obj.len = parseFloat(match[1]);
        // eslint-disable-next-line prefer-destructuring
        obj.unit = match[2];
    }

    return obj;
}

/**
 * @param {HTMLElement} _elem
 * @param {object} [_style]
 * @param {boolean} [_toFloat]
 * @return {object}
 */
export function getComputedStyle(_elem, _style, _toFloat) {
    let css = {};
    let cStyle;
    let val;
    let propVal;

    if (_elem) {
        cStyle = window.getComputedStyle(_elem);

        if (isObject(_style)) {
            for (const prop in _style) {
                if (prop in cStyle || (propVal = cStyle.getPropertyValue(prop)) !== '') {
                    // console.log('???', prop, prop.indexOf('--'));
                    if (propVal !== '') {
                        // css custom property
                        val = propVal;
                        propVal = '';
                    } else {
                        val = cStyle[prop];
                    }
                    // _style[prop] = cStyle[toKebabCase(prop)];
                    _style[prop] = val;

                    if (_toFloat) {
                        _style[prop] = parseFloat(_style[prop]);
                    }
                }
            }

            css = _style;
        } else {
            css = cStyle;
        }
    }

    return css;
}
