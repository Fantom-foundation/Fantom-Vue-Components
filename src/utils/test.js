/**
 * Utility functions for testing
 */

import { toKebabCase } from './string.js';

/**
 * @param {function} func Function to be called
 * @param {int} [n] Number of calls
 * @param {int} [interval] Time span between two calls in milliseconds
 * @return {Promise<unknown>}
 */
export function callFunctionNTimes(func, n = 3, interval = 25) {
    return new Promise(resolve => {
        let counter = n;
        const intervalId = setInterval(() => {
            if (counter-- <= 0) {
                clearInterval(intervalId);
                resolve();
            } else {
                func();
            }
        }, interval);
    });
}

/**
 * Disable console error messages.
 *
 * @param {function} func
 */
export function disableErrorMessages(func) {
    const originalConsoleError = console.error;
    console.error = () => {};

    func();

    console.error = originalConsoleError;
}

/**
 * Playground for testing v-model of a Vue component.
 *
 * @param {Vue} component
 * @param {Object} data
 * @return {{dValue}|{template: `<${string} v-model="dValue" ${string} />`, components: {}, data(): {dValue: *}, props: string[]}}
 */
export function getVModelComponent(component, data = {}) {
    const componentName = component.name;
    let attributes = [];

    Object.keys(data).forEach(key => {
        attributes.push(`:${key}="${key}"`);
    });

    return {
        components: { [componentName]: component },
        template: `<${toKebabCase(componentName)} v-model="dValue" ${attributes.join(' ')} />`,
        props: ['value'],
        data() {
            return {
                dValue: this.value,
                ...data,
            };
        },
    };
}

/**
 * @param {Object} wrapper
 */
export function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}
