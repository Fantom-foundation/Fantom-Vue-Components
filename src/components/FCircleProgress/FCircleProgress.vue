<template>
    <div class="fcircleprogress" :class="classes">
        <svg viewBox="-50 -50 100 100">
            <circle :r="r" cx="0" cy="0" fill="none" :stroke-width="strokeWidth" class="fcircleprogress_bgcircle" />
            <circle
                v-if="dashoffset < 1"
                ref="circle"
                :r="r"
                cx="0"
                cy="0"
                fill="none"
                :stroke-width="strokeWidth"
                stroke-dasharray="1"
                :stroke-dashoffset="dashoffset"
                transform="rotate(-90 0 0)"
                pathLength="1"
                class="fcircleprogress_circle"
            />
            <circle
                v-for="(dot, index) in dots"
                :key="dot[idSymbol]"
                :r="strokeWidth / 3"
                :cx="dotCx(index)"
                :cy="dotCy(index)"
                :fill="dot.color"
                class="fcircleprogress_dot"
            />
            <text text-anchor="middle" dominant-baseline="central" class="fcircleprogress_value" :style="textStyle">
                <slot :value="cValue" :percentage="percentage" :convertToPercentages="convertToPercentages">
                    <slot
                        name="value"
                        :value="cValue"
                        :percentage="percentage"
                        :convertToPercentages="convertToPercentages"
                    >
                        <tspan>{{ convertToPercentages ? percentage : cValue }}</tspan>
                    </slot>
                    <tspan
                        v-if="convertToPercentages"
                        dominant-baseline="central"
                        class="fcircleprogress_value_percentage"
                        dx="-5"
                    >
                        %
                    </tspan>
                </slot>
            </text>
        </svg>
    </div>
</template>

<script>
import { clamp } from '../../utils/number.js';
import { setCustomProperty } from '../../utils/css.js';
import { setIds } from '../../utils/vue-helpers.js';

/**
 *
 */
export default {
    name: 'FCircleProgress',

    props: {
        /** Current value. */
        value: {
            type: [Number, String],
            default: 10,
        },
        /** Minimal value. */
        min: {
            type: Number,
            default: 0,
        },
        /** Maximal value. */
        max: {
            type: Number,
            default: 100,
        },
        /**
         * Css styles of text and circle's color for given values.
         * If current value is bigger than value from `styles` array, css styles from array will be used.
         * @type {{textStyle: Object, color: string, value: number}[]}
         */
        styles: {
            type: Array,
            default() {
                return [];
            },
        },
        /**
         * Dots displayed on circle at positions given by values.
         * @type {{color: string, value: number}[]}
         */
        dots: {
            type: Array,
            default() {
                return [];
            },
        },
        /** Width of circle stroke. */
        strokeWidth: {
            type: Number,
            default: 5,
        },
        /** Animate progress circle. */
        animate: {
            type: Boolean,
            default: false,
        },
        /** Show percentage rather than value. */
        convertToPercentages: {
            type: Boolean,
            default: false,
        },
        /** Can exceed `max` value. */
        canExceed: {
            type: Boolean,
            default: false,
        },
    },

    data() {
        return {
            r: 50 - this.strokeWidth / 2,
            idSymbol: Symbol(),
        };
    },

    computed: {
        cValue() {
            let value = this.value;

            value = clamp(value, this.min, this.max);

            if (this.canExceed && value === this.max) {
                value = this.value;
            }

            this.setCircleStrokeColor(value);

            return value;
        },

        percentage() {
            return this.getPercentage(this.cValue);
        },

        textStyle() {
            return this.getStyleByValue(this.cValue, this.styles, 'textStyle');
        },

        dashoffset() {
            const percentage = this.percentage > 100 ? 100 : this.percentage;

            return 1 - percentage / 100;
        },

        classes() {
            return {
                'fcircleprogress-animate': this.animate,
            };
        },
    },

    created() {
        if (this.max < this.min) {
            throw new Error("'max' must be bigger than 'min'");
        }

        setIds(this.dots, this.idSymbol);
    },

    mounted() {
        this.setCircleStrokeColor(this.cValue);
    },

    methods: {
        /**
         * @param {number} _index Index into `dots` array.
         */
        dotCx(_index) {
            return Math.cos(this.getDotX(_index)) * this.r;
        },

        /**
         * @param {number} _index Index into `dots` array.
         */
        dotCy(_index) {
            return Math.sin(this.getDotX(_index)) * this.r;
        },

        /**
         * @param {number} _index Index into `dots` array.
         */
        getDotX(_index) {
            const PI = Math.PI;
            let percentage = this.getPercentage(this.dots[_index].value);

            if (percentage > 100) {
                percentage = 100;
            }

            return 2 * PI * (percentage / 100) - PI / 2;
        },

        /**
         * @param {number} value
         */
        getPercentage(value) {
            return Math.round(((value - this.min) / (this.max - this.min)) * 100);
        },

        /**
         * @param {number} value
         * @param {array} styles
         * @param {string} prop
         */
        getStyleByValue(value, styles, prop) {
            let style = {};

            for (let i = styles.length - 1; i >= 0; i--) {
                if (value >= styles[i].value) {
                    if (prop === 'color') {
                        style = { color: styles[i].color };
                    } else {
                        style = styles[i][prop];
                    }
                    break;
                }
            }

            return style;
        },

        /**
         * Set circle stroke color according to current value and `this.colors` array.
         *
         * @param {number} value
         */
        setCircleStrokeColor(value) {
            const { $el } = this;
            const { styles } = this;

            if (!$el || styles.length === 0) {
                return;
            }

            const style = this.getStyleByValue(value, styles, 'color');

            if (style.color) {
                setCustomProperty('--fcircleprogress-circle-color', style.color, $el);
            } else {
                setCustomProperty('--fcircleprogress-circle-color', '', $el);
            }
        },
    },
};
</script>

<style lang="scss">
@use 'style';
</style>
