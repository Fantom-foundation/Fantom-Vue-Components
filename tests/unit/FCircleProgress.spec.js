import { mount } from '@vue/test-utils';
import FCircleProgress from '@/components/FCircleProgress/FCircleProgress.vue';
import { disableErrorMessages } from '@/utils/test.js';
import { getCustomProperty } from '../../src/utils/css.js';

let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper({ propsData = { value: 10, min: 0, max: 100 }, slots = {} } = {}) {
    return mount(FCircleProgress, {
        propsData,
        slots,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FCircleProgress', () => {
    it('should render correctly', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should throw an error if `min` is greater than `max`', () => {
        disableErrorMessages(function() {
            expect(() => {
                createWrapper({
                    propsData: {
                        min: 100,
                        max: 50,
                    },
                });
            }).toThrow();
        });
    });

    it('should clamp given value', async () => {
        wrapper = createWrapper({
            propsData: {
                value: -10,
            },
        });

        expect(wrapper.vm.cValue).toBe(0);

        await wrapper.setProps({ value: 200 });

        expect(wrapper.vm.cValue).toBe(100);
    });

    it('should not clamp max value if `canExceed` is set to `true`', async () => {
        wrapper = createWrapper({
            propsData: {
                value: -10,
                canExceed: true,
            },
        });

        expect(wrapper.vm.cValue).toBe(0);

        await wrapper.setProps({ value: 200 });

        expect(wrapper.vm.cValue).toBe(200);
    });

    it('should convert given value to percentages if `convertToPercentages` prop is `true`', () => {
        wrapper = createWrapper({
            propsData: {
                value: 60,
                min: 20,
                max: 100,
                convertToPercentages: true,
            },
        });

        expect(
            wrapper
                .find('.fcircleprogress_value')
                .text()
                .replace(/[ \n\r\t]/g, '')
        ).toBe('50%');
    });

    it('should apply css style to the text element if value is within given range and `styles` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                styles: [
                    {
                        value: 0,
                        textStyle: { fill: 'red' },
                    },
                    {
                        value: 40,
                        textStyle: { fill: 'orange' },
                    },
                    {
                        value: 80,
                        textStyle: { fill: 'green' },
                    },
                ],
            },
        });

        const elem = wrapper.find('.fcircleprogress_value').element;

        expect(elem.style.fill).toBe('red');

        await wrapper.setProps({ value: 41 });
        expect(elem.style.fill).toBe('orange');

        await wrapper.setProps({ value: 80 });
        expect(elem.style.fill).toBe('green');
    });

    it('should apply color to the circle if value is within given range and `styles` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                styles: [
                    {
                        value: 0,
                        color: 'red',
                    },
                    {
                        value: 40,
                        color: 'orange',
                    },
                    {
                        value: 80,
                        color: 'green',
                    },
                ],
            },
        });

        const elem = wrapper.element;

        expect(getCustomProperty('--fcircleprogress-circle-color', null, elem)).toBe('red');

        await wrapper.setProps({ value: 41 });
        expect(getCustomProperty('--fcircleprogress-circle-color', null, elem)).toBe('orange');

        await wrapper.setProps({ value: 80 });
        expect(getCustomProperty('--fcircleprogress-circle-color', null, elem)).toBe('green');
    });

    it('should draw colored dots at given value if `dots` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                dots: [
                    {
                        value: 30,
                        color: 'green',
                    },
                    {
                        value: 75,
                        color: 'red',
                    },
                ],
            },
        });

        const eDots = wrapper.findAll('.fcircleprogress_dot');

        expect(eDots.length).toBe(2);
        expect(eDots.at(0).attributes('fill')).toBe('green');
        expect(eDots.at(1).attributes('fill')).toBe('red');
    });

    it('should set `stroke-width` attribute correctly if `strokeWidth` prop is set', async () => {
        wrapper = createWrapper({
            propsData: {
                strokeWidth: 10,
            },
        });

        expect(wrapper.find('.fcircleprogress_circle').attributes('stroke-width')).toBe('10');
    });
});
