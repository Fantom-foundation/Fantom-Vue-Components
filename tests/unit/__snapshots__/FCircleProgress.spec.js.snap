// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FCircleProgress should render correctly 1`] = `
<div
  class="fcircleprogress"
>
  <svg
    viewBox="-50 -50 100 100"
  >
    <circle
      class="fcircleprogress_bgcircle"
      cx="0"
      cy="0"
      fill="none"
      r="47.5"
      stroke-width="5"
    />
     
    <circle
      class="fcircleprogress_circle"
      cx="0"
      cy="0"
      fill="none"
      pathLength="1"
      r="47.5"
      stroke-dasharray="1"
      stroke-dashoffset="0.9"
      stroke-width="5"
      transform="rotate(-90 0 0)"
    />
      
    <text
      class="fcircleprogress_value"
      dominant-baseline="central"
      text-anchor="middle"
    >
      <tspan>
        10
      </tspan>
       
      <!---->
    </text>
  </svg>
</div>
`;
